### ISSUE TITLE
get_params and set_params for MLClassifierBase erroneous 

### ISSUE THREAD

**Issue by ChristianSch** [scikit-multilearn/scikit-multilearn#29]
Hi,

While writing the test case for cross validation I noticed that I missed something in my implementation of said functions, which gives me time to discuss some shortcomings of the code.

In the scikit-learn implementation, there is quite a lot of code to find out the attributes of each object for the get_param function, namely `_get_param_names` and `signature`. I think for simple cases like the meta estimators we have here this would be just overhead. Thus I'd advocate for a list of attributes to be copied, defined by the very estimators. Or at least the MLClassifierBase could define the common ones, like `copyable_attrs = ["require_dense", "classifier"]`, and if needed, the heiring classifiers would rewrite this attribute with some of their own attributes, or add them or whatever.

In `get_params` we can then run over the keys and copy them, if deep-copyable and `deep=True` copy them too. Also we can check for attributes in `set_params`, providing another safety measure to prevent writing attribues to the object that weren't set in the first place. (This could result in some nasty bugs I think.)

What is your take on this?

PS: I'll commit the bug fixes. Additionally I've removed the get_params/set_params from the label powerset meta estimator. They were wrong anyways (missing `self` as first attribute).

**Comment by niedakh** [comment]
Hi,

i think you are correct, what we would also need is to come up with a way to pass the parameters up/down across the classifier tree, for example lets say we have an ensemble of random ordered classifier chains. EnsembleML(RandomizeLabels(CC(DecisionTreeClassifier))).

SVC takes some params, Randomize Labels might take a seed, EnsembleML will surely take the no of chains and the size of input space subset. 

Do we pass the arguments in a namespaces approach flat array, ex. the get_params from SVC would yield:
```
tree_parms = {'class_weight': None,
 'criterion': 'gini',
 'max_depth': None,
 'max_features': None,
 'max_leaf_nodes': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'presort': False,
 'random_state': None,
 'splitter': 'best'}
```

What does CC return? Do we stack this, like:
```
{ "cc__baseclassifierparams" : self.classifier.get_params(),
"cc__classifier": DecisionTreeClassifier(),
"cc_require_dense": True}
```

or do we make it a flat list
```
{ "cc__classifier": DecisionTreeClassifier(),
"cc_require_dense": True,
"decisiontre...classifier_criterion": ...,
...
"decisiontre...classifier_splitter": ...
}
```

Which one would you prefer? 

I want to cleanup the MLkNN code and convert it to sparse matrices internally and that will finish the 0.0.2 feature list, then I'll look into your code ok?

**Comment by ChristianSch** [comment]
I'd look into how scikit-learn does it and then comply to this. Not sure if they have this nested params, but I'll look into it.

**Comment by ChristianSch** [comment]
Ah, I know. It's flat because the way we iterate the attributes. So flat is the way to go imho.

**Comment by niedakh** [comment]
fully merged as tests in 477efdd012f9696468ce7037dec5967b2b3de2de, thank you again @ChristianSch !
